### PR TITLE
Fix #1000 : Update the color of status bar in various screens

### LINKED ISSUE #1000: Update the color of status bar in various screens
The status bar color in various android screens is inconsistent. Update status bar color code as per the mocks and spreadsheet:
https://docs.google.com/spreadsheets/d/1XphP6nWoUdOZn9pY2kdgfqMWfJjMfKMb98MXM5rp2HY/edit#gid=0
https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/

### PR DESCRIPTION
**Pull request by SayantanBanerjee16**
## Explanation

Fixes #1000 :
Step1) Assigned colors from spreadsheets to color.xml
Step2) Created a Utility method where the update of Status Bar color take place
Step3) Called that Utility method from various places as required.

## Spreadsheet :
https://docs.google.com/spreadsheets/d/1XphP6nWoUdOZn9pY2kdgfqMWfJjMfKMb98MXM5rp2HY/edit#gid=0
https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/ 

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by SayantanBanerjee16** [comment]
Screenshots (In order of appearance in spread-sheet): 

1) Profile Chooser

profilechooser

2) Profile Image

profileimage

3) General Header

generalheader

4) Slider open

draweropen

5) PIN Input

PinInput

6) Walkthrough

walkthrough


7) Onboarding 1 to 4

onboarding1

onboarding2

onboarding3

onboarding4

**Comment by SayantanBanerjee16** [comment]
@rt4914 @nikitamarysolomanpvt  
I am stuck on these files of the spread-sheet in locating them.

```
- Hint Full Dialog
- Concept Card
- Registration
```
Can you please help me out. Thanks :)

File: app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt [hunk 1]
```diff
@@ -55,10 +56,20 @@ class OnboardingFragmentPresenter @Inject constructor(
...
       }
     })
   }
 
+  private fun onboardingStatusBarColorUpdate(position: Int) {
+    when (position) {
+      0 -> StatusBarColor.statusBarColorUpdate(R.color.onboarding_1_status_bar, activity, false)
+      1 -> StatusBarColor.statusBarColorUpdate(R.color.onboarding_2_status_bar, activity, false)
+      2 -> StatusBarColor.statusBarColorUpdate(R.color.onboarding_3_status_bar, activity, false)
+      3 -> StatusBarColor.statusBarColorUpdate(R.color.onboarding_4_status_bar, activity, false)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt]: Add an else clause also, which can point to `#003933` as default.

File: app/src/main/res/values/colors.xml [hunk 2]
```diff
@@ -135,4 +135,16 @@
...
+  <color name="hint_full_dialog_status_bar">#00236E</color>
+  <color name="slide_drawer_open_status_bar">#001F2A</color>
+  <color name="concept_card_status_bar">#882814</color>
+  <color name="pin_input_status_bar">#BDCCCC</color>
+  <color name="walkthrough_status_bar">#CCCCCC</color>
+  <color name="onboarding_1_status_bar">#00847A</color>
+  <color name="onboarding_2_status_bar">#8B2A15</color>
+  <color name="onboarding_3_status_bar">#AD8500</color>
+  <color name="onboarding_4_status_bar">#3B5893</color>
+  <color name="registration_status_bar">#9CB1BE</color>
```
Comment (rt4914) [inline app/src/main/res/values/colors.xml]: Rename all the colors introduced by you in camelCase format:
example:
`onboarding1StatusBar `

File: utility/src/main/java/org/oppia/util/statusbar/StatusBarColor.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package org.oppia.util.statusbar
+
+import android.os.Build
+import android.view.View
+import androidx.appcompat.app.AppCompatActivity
+import androidx.core.content.ContextCompat
+
+class StatusBarColor {
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/statusbar/StatusBarColor.kt:9]: Add a Kdoc for this class explaining what it does.

File: utility/src/main/java/org/oppia/util/statusbar/StatusBarColor.kt [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  companion object {
+    fun statusBarColorUpdate(colorId: Int, activity: AppCompatActivity, statusBarLight: Boolean) {
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/statusbar/StatusBarColor.kt:17]: Add a kdoc for this functions mentioning all its params and it does.

**Review by rt4914 — CHANGES_REQUESTED** [review]
This is very nice implementation. I have suggested some changes. Also, HintsAnd Solution is missing this change.

Screenshot_1588048991


For viewing hints follow these steps:

1. Start `What is Fraction?` exploration.
2. Click on `Continue` till you reach your first question.
3. In first question, select first option twice as it would be an incorrect answer. wait for 10 secs and in you will see blue box on bottom-left, open it, this screen is for HintsAndSolution.

Thanks.

**Comment by rt4914** [comment]
> @rt4914 @nikitamarysolomanpvt
> I am stuck on these files of the spread-sheet in locating them.
> 
> ```
> - Hint Full Dialog
> - Concept Card
> - Registration
> ```
> 
> Can you please help me out. Thanks :)

I have replied above for Hints.
For concept card you can make `ContentCardTestActivity` as launcher activity and test but I think it already has a color.
`Registration` feature is not available yet, so you can leave it.

**Comment by nikitamarysolomanpvt** [comment]
> ## Explanation
> Fixes #1000 :
> Step1) Assigned colors from spreadsheets to color.xml
> Step2) Created a Utility method where the update of Status Bar color take place
> Step3) Called that Utility method from various places as required.
> 
> ## Checklist
> * [x]  The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
> * [x]  The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
> * [x]  The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
> * [x]  The PR does not contain any unnecessary auto-generated code from Android Studio.
> * [x]  The PR is made from a branch that's **not** called "develop".
> * [x]  The PR is made from a branch that is up-to-date with "develop".
> * [x]  The PR's branch is based on "develop" and not on any other branch.
> * [x]  The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

@SayantanBanerjee16 Please mention the mock link or sheet link in PR for review.

File: app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt (hunk 1 above)
Comment (nikitamarysolomanpvt) [reply app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt]: +1

Comment (SayantanBanerjee16) [reply app/src/main/java/org/oppia/app/onboarding/OnboardingFragmentPresenter.kt]: Done

File: app/src/main/res/values/colors.xml (hunk 2 above)
Comment (SayantanBanerjee16) [reply app/src/main/res/values/colors.xml]: Done

File: utility/src/main/java/org/oppia/util/statusbar/StatusBarColor.kt (hunk 3 above)
Comment (nikitamarysolomanpvt) [reply utility/src/main/java/org/oppia/util/statusbar/StatusBarColor.kt:9]: +1

Comment (SayantanBanerjee16) [reply utility/src/main/java/org/oppia/util/statusbar/StatusBarColor.kt:9]: Done

File: utility/src/main/java/org/oppia/util/statusbar/StatusBarColor.kt (hunk 4 above)
Comment (SayantanBanerjee16) [reply utility/src/main/java/org/oppia/util/statusbar/StatusBarColor.kt:17]: Done

File: utility/src/main/java/org/oppia/util/statusbar/StatusBarColor.kt (hunk 3 above)
Comment (nikitamarysolomanpvt) [reply utility/src/main/java/org/oppia/util/statusbar/StatusBarColor.kt:8]: +1

File: utility/src/main/java/org/oppia/util/statusbar/StatusBarColor.kt (hunk 4 above)
Comment (nikitamarysolomanpvt) [reply utility/src/main/java/org/oppia/util/statusbar/StatusBarColor.kt:10]: +1

**Review by nikitamarysolomanpvt — CHANGES_REQUESTED** [review]
PTAL on comments of @rt4914.
Please include Kdocs
@rt4914 we may use statusbarcolor in in toolbar view model as well, am waiting for ben's word.

**Comment by SayantanBanerjee16** [comment]
> PTAL on comments of @rt4914.
> Please include Kdocs
> @rt4914 we may use statusbarcolor in in toolbar view model as well, am waiting for ben's word.

@rt4914 @nikitamarysolomanpvt Addressed all the comments. Introduced camelCase colorID and Kdoc also.

Hints and Solution Dialogs Screenshot:-
hints

PTAL. Thanks :)

**Comment by nikitamarysolomanpvt** [comment]
> > PTAL on comments of @rt4914.
> > Please include Kdocs
> > @rt4914 we may use statusbarcolor in in toolbar view model as well, am waiting for ben's word.
> 
> @rt4914 @nikitamarysolomanpvt Addressed all the comments. Introduced camelCase colorID and Kdoc also.
> 
> Hints and Solution Dialogs Screenshot:-
> hints
> 
> PTAL. Thanks :)

@SayantanBanerjee16  Please click resolve conversation after addressing comments.

**Comment by SayantanBanerjee16** [comment]
> > > PTAL on comments of @rt4914.
> > > Please include Kdocs
> > > @rt4914 we may use statusbarcolor in in toolbar view model as well, am waiting for ben's word.
> > 
> > 
> > @rt4914 @nikitamarysolomanpvt Addressed all the comments. Introduced camelCase colorID and Kdoc also.
> > Hints and Solution Dialogs Screenshot:-
> > hints
> > PTAL. Thanks :)
> 
> @SayantanBanerjee16 Please click resolve conversation after addressing comments.

@nikitamarysolomanpvt All the pending conversations are Done and resolved.

**Review by nikitamarysolomanpvt — CHANGES_REQUESTED** [review]
Please check it in kitkat seems not working.

**Comment by rt4914** [comment]
> Please check it in kitkat seems not working.

@SayantanBanerjee16 Check this in kitkat and see how much work do we need to fix this. If it can be fixed in this PR itself then its great.

**Comment by SayantanBanerjee16** [comment]
> > Please check it in kitkat seems not working.
> 
> @SayantanBanerjee16 Check this in kitkat and see how much work do we need to fix this. If it can be fixed in this PR itself then its great.

@rt4914 @nikitamarysolomanpvt 

So change of status bar color can be done on versions above LOLLIPOP and application of dark icon themes can be done from above MARSHMELLO.

For KITKAT, I approached a method from a [stack overflow answer](https://stackoverflow.com/questions/34042232/set-status-bar-color-dynamically-in-android/34048393#34048393) using a [github library](https://github.com/jgilfelt/SystemBarTint) but although it does draw color to system bar, it is getting overlapped with other screen contents.
Shall I revert this KITKAT library code?

Screenshot (115)

Screenshot (116)

File: app/src/main/res/layout-land/walkthrough_final_fragment.xml [hunk 5]
```diff
@@ -23,7 +23,7 @@
       android:layout_width="180dp"
       android:layout_height="180dp"
       android:layout_marginStart="28dp"
-      android:src="@drawable/ic_onboarding_0"
+      android:src="@drawable/ic_landscape_onboarding_0"
```
Comment (SayantanBanerjee16) [inline app/src/main/res/layout-land/walkthrough_final_fragment.xml]: It is causing build failure in my system saying resource `ic_onboarding_0` doesn't exist. Should I revert them?

File: app/src/main/res/layout-land/walkthrough_welcome_fragment.xml [hunk 6]
```diff
@@ -52,7 +52,7 @@
       android:layout_width="180dp"
       android:layout_height="180dp"
       android:layout_marginStart="56dp"
-      android:src="@drawable/ic_onboarding_0"
+      android:src="@drawable/ic_landscape_onboarding_0"
```
Comment (SayantanBanerjee16) [inline app/src/main/res/layout-land/walkthrough_welcome_fragment.xml]: Ditto as above

Comment (nikitamarysolomanpvt) [reply app/src/main/res/layout-land/walkthrough_welcome_fragment.xml]: @SayantanBanerjee16 
will create a PR to fix this

Comment (SayantanBanerjee16) [reply app/src/main/res/layout-land/walkthrough_welcome_fragment.xml]: @nikitamarysolomanpvt Okay am reverting it from this PR.

Comment (nikitamarysolomanpvt) [reply app/src/main/res/layout-land/walkthrough_welcome_fragment.xml]: @SayantanBanerjee16 You may update your code with develop now.

Comment (SayantanBanerjee16) [reply app/src/main/res/layout-land/walkthrough_welcome_fragment.xml]: Reverted

File: app/src/main/res/layout-land/walkthrough_final_fragment.xml (hunk 5 above)
Comment (SayantanBanerjee16) [reply app/src/main/res/layout-land/walkthrough_final_fragment.xml]: Reverted.

**Comment by rt4914** [comment]
> > > Please check it in kitkat seems not working.
> > 
> > 
> > @SayantanBanerjee16 Check this in kitkat and see how much work do we need to fix this. If it can be fixed in this PR itself then its great.
> 
> @rt4914 @nikitamarysolomanpvt
> 
> So change of status bar color can be done on versions above LOLLIPOP and application of dark icon themes can be done from above MARSHMELLO.
> 
> For KITKAT, I approached a method from a [stack overflow answer](https://stackoverflow.com/questions/34042232/set-status-bar-color-dynamically-in-android/34048393#34048393) using a [github library](https://github.com/jgilfelt/SystemBarTint) but although it does draw color to system bar, it is getting overlapped with other screen contents.
> Shall I revert this KITKAT library code?
> 
> Screenshot (115)
> 
> Screenshot (116)

Yes, this solution is not appropriate as it is overlapping on other views. You should try some other solutions for this.

**Comment by SayantanBanerjee16** [comment]
@rt4914 Sorry but am unable to find any solution linked to applying color to status bar in Pre- Lollipop versions. I have reverted the previous introduced code for KITKAT.

**Comment by rt4914** [comment]
> @rt4914 Sorry but am unable to find any solution linked to applying color to status bar in Pre- Lollipop versions. I have reverted the previous introduced code for KITKAT.

@nikitamarysolomanpvt PTAL. Thanks.

**Comment by nikitamarysolomanpvt** [comment]
> @rt4914 Sorry but am unable to find any solution linked to applying color to status bar in Pre- Lollipop versions. I have reverted the previous introduced code for KITKAT.

@SayantanBanerjee16 try https://stackoverflow.com/questions/22192291/how-to-change-the-status-bar-color-in-android/26215844#26215844
You may merge this and try as another PR. @rt4914 Please guide him with the same.

**Review by nikitamarysolomanpvt — APPROVED** [review]
LGTM except the pre lollipop (Kitkat) issue

**Review by rt4914 — APPROVED** [review]
LGTM, also I have filed an issue to track this. #1042

**Pull request merged**